File: Python/find_far_field.py
import numpy as np
def unit_normal(k, n, x, dx, ddx, f,eta,phi,OP,t,g,R,A,angles):
    v = dx(t)[:,:,0].T@np.array(((0,-1),(1,0)))
    r = np.sqrt(np.sum(v**2,axis=1))
    r = r.reshape(len(r),1)
    v = v/np.tile(r,(1,2))
    return v
def find_far_field(k, n, x, dx, ddx, f,eta,phi,OP,t,g,R,A,angles):
    angles = angles.reshape(1,len(angles))
    t = t.reshape(len(t),1)
    y = x(t)[:,:,0].T
    dy = dx(t)[:,:,0].T
    ds = np.sqrt(np.sum(dy**2,axis=1))
    phi = phi.reshape(phi.shape[0])
    xhat = np.concatenate((np.cos(angles),np.sin(angles)),axis=0)
    nu = unit_normal(k, n, x, dx, ddx, f,eta,phi,OP,t,g,R,A,angles)
    uinf = np.zeros((angles.shape[1],1))+1j*np.zeros((angles.shape[1],1))
    a = np.exp(-1j*np.pi/4)/np.sqrt(8*np.pi*k)
    for j in range(xhat.shape[1]):
        xh = xhat[:,j]
        K = (k*nu@xh+eta)*np.exp(-1j*k*y@xh)*phi*ds
        uinf[j] = a*np.sum(K*np.pi/n,axis=0)
    return uinf,nu,y,xhat,phi

File: Matlab/find_far_field.m
%反散射那本书96页远场的公式
function [uinf, nu, y, xhat, phi] = find_far_field(obj, sol, angles)
    N = obj.n;
    x   = obj.x;
    dx  = obj.dx;
    k   = obj.k;
    t = (0:2*N-1)'*pi/N;
    y = x(t);
    dy = dx(t);
    ds = sqrt(sum(dy.^2, 2));
    s = size(angles);
    if s(1) > s(2)
        angles = angles';
    end
    xhat = [cos(angles); sin(angles)];
    nu = unit_normal(obj, t);
    eta = sol.eta;
    phi = sol.phi;
    uinf = zeros(size(angles'));
    a = exp(-1i*pi/4) / sqrt(8*pi*k);
    for j = 1:length(xhat)
        xh = xhat(:,j);
        K = (k*nu*xh + eta) .* exp(-1i*k*y*xh) .* phi .* ds; 
        uinf(j) = a * sum(K * pi/N);
    end
end


function v = unit_normal(obj, t)
    dx = obj.dx;
    s = size(t);
    if s(2) > s(1)
        t = t';
    end
    v = dx(t) * [0 -1; 1 0];
    r = sqrt(sum(v.^2, 2));
    v = v ./ repmat(r, [1 2]);
end 